### PR TITLE
Add basic support for books in `tes3.createObject`

### PR DESCRIPTION
**Pull request by Virnetch**
Notes:
 - No support for adding text to the books. This would require additional work since morrowind doesn't seem to be able to read the TEXT subrecord from save files.
 - `enchantCapacity` of unenchanted scrolls must be zero. The game seems to crash when trying to enchant books that are from the save game so I prevented creation of scrolls with an enchantCapacity.

I don't have the skills to fix these myself. This is also mostly copy-pasted from other places but I haven't encountered any other issues with it so far. Also, this is my first ever pull request so sorry if there's something I did wrong 😄 

There was an earlier PR that also added some kind of support for saving text here: #255

### CODE REVIEW

File: MWSE/LuaObject.h [hunk 1]
```diff
@@ -576,6 +577,81 @@ namespace mwse::lua {
 		}
 	};
 
+	template<>
+	class ObjectCreator<TES3::Book> : public ObjectCreatorBase {
+	public:
+		TES3::BaseObject* create(sol::table params, bool getIfExists) const override {
```
Comment (pierresteeg) [inline MWSE/LuaObject.h:583]: `getIfExists` does not get checked.

File: MWSE/LuaObject.h [hunk 2]
```diff
@@ -576,6 +577,81 @@ namespace mwse::lua {
...
+			auto enchantment = getOptionalParamObject<TES3::Enchantment>(params, "enchantment");
+			auto enchantCapacity = getOptionalParam<unsigned short>(params, "enchantCapacity", 0);
+			auto bookType = getOptionalParam<unsigned char>(params, "type", 0);
+			if (enchantment && enchantment->objectType == TES3::ObjectType::Enchantment) {
+				book->enchantment = enchantment;
+			}
+			else if (enchantCapacity > 0 && bookType == 1)
+				// Prevent creation of books that the player could enchant.
+				// The game crashes if trying to enchant books that aren't from an esp/esm plugin.
+				throw std::invalid_argument{ "tes3.createObject: 'enchantCapacity' parameter must be zero if creating an unenchanted scroll (the game will crash if player tries to enchant it)." };
```
Comment (pierresteeg) [inline MWSE/LuaObject.h:627]: Something minor, I'd move the closing bracket to a new line here.

File: MWSE/LuaObject.h [hunk 3]
```diff
@@ -576,6 +577,81 @@ namespace mwse::lua {
...
+			if (script != nullptr)
+				book->script = script;
+
+			auto enchantment = getOptionalParamObject<TES3::Enchantment>(params, "enchantment");
+			auto enchantCapacity = getOptionalParam<unsigned short>(params, "enchantCapacity", 0);
+			auto bookType = getOptionalParam<unsigned char>(params, "type", 0);
+			if (enchantment && enchantment->objectType == TES3::ObjectType::Enchantment) {
+				book->enchantment = enchantment;
+			}
+			else if (enchantCapacity > 0 && bookType == 1)
```
Comment (pierresteeg) [inline MWSE/LuaObject.h:624]: You probably want to create an enum for this one in TES3Book.h. You could see `TES3::SpellEffectState` for an example.

File: MWSE/LuaObject.h (hunk 1 above)
Comment (Virnetch) [reply MWSE/LuaObject.h:583]: Looks like `enchantment` has the same issue that I copied over...

Comment (pierresteeg) [reply MWSE/LuaObject.h:583]: That is correct. Ill fix that for the enchantment, thanks.

File: MWSE/LuaObject.cpp [hunk 4]
```diff
@@ -23,6 +23,8 @@ namespace mwse::lua {
 			return std::make_unique<ObjectCreator<TES3::Enchantment>>();
 		case TES3::ObjectType::Weapon:
 			return std::make_unique<ObjectCreator<TES3::Weapon>>();
+		case TES3::ObjectType::Book:
+			return std::make_unique<ObjectCreator<TES3::Book>>();
 		default:
```
Comment (C3pa) [inline MWSE/LuaObject.cpp:28]: This isn't necessary, but what about keeping the switch case sorted alphabetically, and adding this below `TES3::ObjectType::Ammo`?

File: MWSE/LuaObject.h (hunk 3 above)
Comment (Virnetch) [reply MWSE/LuaObject.h]: Done. Let me know if there's any issues with it.

File: MWSE/LuaObject.cpp (hunk 4 above)
Comment (Virnetch) [reply MWSE/LuaObject.cpp:28]: Yeah I'll move it there. Enchantment should probably also be moved.

File: MWSE/LuaObject.h (hunk 3 above)
Comment (pierresteeg) [reply MWSE/LuaObject.h]: Looks fine to me, thanks!

**Comment by Virnetch** [comment]
@NullCascade When you have the time, could you take a look at this? I think it should be fine to be merged, unless someone wants to add support for the `text` field or fix the crashes when enchanting.

File: MWSE/LuaObject.h (hunk 2 above)
Comment (pierresteeg) [reply MWSE/LuaObject.h]: Whoops, totally looked at the wrong line here, sorry! Ignore the comment.

**Pull request merged**